### PR TITLE
Add notion of "default" bindings

### LINKED ISSUE #3699: fish_vi_key_bindings doesn't call fish_user_key_bindings
- [x] Have you checked if problem occurs with [fish 2.4.0](/fish-shell/fish-shell/releases/tag/2.4.0)?
- [x] Tried fish without third-party customizations *(check `sh -c 'env HOME=$(mktemp -d) fish'`)*?

**fish version installed** *(`fish --version`)*: 2.4.0

**OS/terminal used**: macOS 10.12.2 iTerm 3.0.13

fish_vi_key_bindings doesn't call fish_user_key_bindings at the end so reloading my config.fish causes my keybindings to disappear. I can manually call `fish_user_key_bindings` in my config.fish but it'd be cool if it was automatic.

### LINKED ISSUE #5191: Modularize user key bindings
Would you consider making fish aware of a new keybindings directory inside `$__fish_datadir`, `XDG_CONFIG_HOME`, etc.?

Then, during startup, copy the user's `fish_user_key_bindings` function (if there is one) and create a new `fish_user_key_bindings` function that when invoked, will <samp>source</samp> every file in each `keybindings/` directory in `$__fish_datadir` and `XDG_CONFIG_HOME`.

Basically this:

```fish
if functions -q fish_user_key_bindings
  functions -c fish_user_key_bindings __fish_user_key_bindings
end

function fish_user_key_bindings
  for file in ~/.config/fish/keybindings/*.fish
    source $file
  end
  if functions -q __fish_user_key_bindings
    __fish_user_key_bindings
  end
end
```

I can emulate this behavior using `conf.d` and distribute that script as a plugin, but it would be nice to have this out of the box.

In fisher v2.x (since I depend exclusively on fish autoloading feature) I go to great lengths to generate a `fish_user_key_bindings` function on the fly that concatenates the keybindings exported by each plugin with the user's key bindings defined inside their `fish_user_key_bindings`, but it's highly impractical (and buggy). In fisher v3.x I'll work around this more cleanly with a plugin as suggested above, but it would be nice if I didn't need to.

Oh-my-fish ([here](https://github.com/oh-my-fish/oh-my-fish/blob/112fb3dbc2b57077913c16b5e992245d278705eb/init.fish#L27-L46)) and fundle ([here](https://github.com/tuvistavie/fundle/blob/d2f688c6dd1df0ea167d47ab6d2c586778437ede/functions/fundle.fish#L247-L260)) emulate this behavior through the user's `config.fish` using the same (or similar) technique described above.

### PR DESCRIPTION
**Pull request by faho**
## Description

As alluded to in #5191, this adds the idea of "default" and "user" bindings to `bind`. A user binding (done with `bind` without any special option) will override a default binding (done with `bind --default`), until it is erased (and the default becomes active again).

What this allows is for us to erase _just_ the default bindings when the user switches the bind set, which allows for `bind` statements in e.g. config.fish to just work.

There's some open questions here, mainly how much we should separate the two binding "levels". I chose to just always require `--default` before showing or doing anything to default bindings, so a bare `bind` will probably show nothing to most people. This was mainly for ease of implementation - I could just add a "bool user" to everything.

Fixes #5191.
Fixes #3699.

## TODOs:

- [x] Changes to fish usage are reflected in user documentation/manpages.
- [x] Tests have been added for regressions fixed
- [x] User-visible changes noted in CHANGELOG.md

### CODE REVIEW

**Comment by zanchey** [comment]
`bind` is a slightly useful way of making current (default) bindings visible, so it might be worth showing default and user bindings with a bare `bind`.

**Comment by faho** [comment]
Okay, so `bind` shows all bindings. `bind --default` presumably shows only the default bindings, so we probably need an explicit "--user" flag now.

Should `bind somekey` show both bindings, or only the active one? The former gives more information, but breaks the assumption that it will only ever print one entry.

**Comment by jorgebucaran** [comment]
I love this. Now I can create packages that add their own key bindings via conf.d. It would work like this:

```
fish-hello
├── functions
│   └── hello.fish
└── conf.d
    └── hello.fish
```
```fish
bind \cs hello

function hello_key_bindings --on-event hello_key_bindings_uninstall
    bind -e \cs hello
end
```

File: CHANGELOG.md [hunk 1]
```diff
@@ -67,6 +67,7 @@ fish 3.0 is a major release which brings with it both improvements in functional
 - `test` and `[` now support floating point values in numeric comparisons.
 - Autosuggestions try to avoid arguments that are already present in the command line.
 - Variables may be used inside commands (#154).
+- Bare `bind` invocations in config.fish now work. The `fish_user_key_bindings` function is no longer necessary, but will still be executed if it exists (#5191).
```
Comment (jorgebucaran) [inline CHANGELOG.md]: @faho Would it be relevant to mention that this is not limited to config.fish, but also to conf.d/*.fish scripts?

Comment (faho) [reply CHANGELOG.md]: The assumption is if something works in config.fish, then it'll also work in conf.d. No need to be that complete here!

**Comment by faho** [comment]
So this is now done.

I'd like to give a final opportunity for bike shedding. The terminology I've used isn't fantastic - the `bind` man page now has three different meanings for the word "default". You can now set the default binding (meaning the binding that is used when nothing else matches) in the default mode (which only really makes sense if you know about vi) and default level (which this is about).

So: Does anyone have a different idea for "default" and "user" bindings? "fallback" and "custom"? Or "standard", "preset"?

And does anyone have an idea for what this concept should be called? Binding "level"?

**Comment by zanchey** [comment]
Wow, default is really overloaded. What about builtin (collides with builtin commands) or basic?

As a name for the feature, what about binding sets?

**Comment by jorgebucaran** [comment]
I think --default makes sense. Perhaps the documentation could be more thorough in explaining the difference between default binding and default level.

**Comment by ridiculousfish** [comment]
This sounds good to me except I share concern about the name "default." In particular the vi key bindings will also be "default" but not in `fish_default_key_bindings` which is bizarre.

I like "system" to distinguish it from user key bindings. `bind --system \n execute`

**Comment by jorgebucaran** [comment]
FWIW I'm more in favor of --preset than --default.

**Comment by faho** [comment]
Merged as 444f9f87154fd3970f11a1bcb8bafee48d2afa76, with the terms "preset" and "user".

Finally!

**Comment by jorgebucaran** [comment]
The default level is `user`, so:

```fish
bind -e \cx "echo hello" # [^+x]
```
and 
```fish
bind --user -e \cx "echo hello" # [^+x]
```
...are equivalent.

**Comment by faho** [comment]
@jorgebucaran: Yes. Adding bindings and erasing bindings defaults to user mode, listing bindings defaults to both.

So

```fish
bind \cx
```

will show both

```fish
bind --preset \cx fish_clipboard_copy
bind \cx "echo hello"
```

erasing is also allowed with both at the same time (so `bind --erase --preset --user` will delete both of those), adding can only do one at a time.

Or, from the docs:

>By default, all `bind` invocations work on the "user" level except for listing, which will show both levels. All invocations except for inserting new bindings can operate on both levels at the same time.

**Comment by ridiculousfish** [comment]
Happy with 'preset'. Awesome work!

**Pull request closed without merging**